src/kernel/arch/x86/arch.zig

```diff
@@ -231,6 +250,80 @@ pub fn haltNoInterrupts() noreturn {
     }
 }
 
+///
+/// Initialise the system's memory. Populates a memory profile with boot modules from grub, the amount of available memory, the reserved regions of virtual and physical memory as well as the start and end of the kernel code
+///
+/// Arguments:
+///     IN mb_info: *multiboot.multiboot_info_t - The multiboot info passed by grub
+///
+/// Return: mem.MemProfile
+///     The constructed memory profile
+///
+/// Error: std.mem.Allocator.Error
+///     std.mem.Allocator.Error.OutOfMemory - There wasn't enough memory in the allocated created to populate the memory profile, consider increasing mem.FIXED_ALLOC_SIZE
+///
+pub fn initMem(mb_info: BootPayload) std.mem.Allocator.Error!MemProfile {
+    log.logInfo("Init mem\n", .{});
+    defer log.logInfo("Done mem\n", .{});
+
+    const mods_count = mb_info.mods_count;
+    mem.ADDR_OFFSET = @ptrToInt(&KERNEL_ADDR_OFFSET);
+    const mmap_addr = mb_info.mmap_addr;
+    const num_mmap_entries = mb_info.mmap_length / @sizeOf(multiboot.multiboot_memory_map_t);
+    const vaddr_end = @ptrCast([*]u8, &KERNEL_VADDR_END);
+
+    var allocator = std.heap.FixedBufferAllocator.init(vaddr_end[0..mem.FIXED_ALLOC_SIZE]);
+    var reserved_physical_mem = std.ArrayList(mem.Range).init(&allocator.allocator);
+    var reserved_virtual_mem = std.ArrayList(mem.Map).init(&allocator.allocator);
+    const mem_map = @intToPtr([*]multiboot.multiboot_memory_map_t, mmap_addr)[0..num_mmap_entries];
+
+    // Reserve the unavailable sections from the multiboot memory map
+    for (mem_map) |entry| {
+        if (entry.@"type" != multiboot.MULTIBOOT_MEMORY_AVAILABLE) {
+            const end: usize = if (entry.addr + entry.len > std.math.maxInt(usize)) std.math.maxInt(usize) else @intCast(usize, entry.addr + entry.len);
```

Comment:
This if can cause a integer overflow, would be better: `entry.addr > std.math.maxInt(usize) - entry.len`. Also, what happens if the end is too long as this would lead to the OS using unavailable memory as it couldn't fit it in a uzise. Maybe best to error here.

Reply:
Nice idea on the if statement. Regarding the other point, what do you mean by the end being too long?

Reply:
if `entry.addr + entry.len > std.math.maxInt(usize)` this is true, then you just use `std.math.maxInt(usize)` so there there will be memory that won't be mapped out.

Reply:
I don't quite see how using `maxInt(usize)` is a problem. Since x86 is 32-bit it can't address the memory beyond `maxInt(usize)` anyway so ignoring whatever comes after that in memory map entry should be fine.

Reply:
Ah yes, good point, maybe add a comment :)

Reply:
Sure thing :)